Sources/Graph.swift

```diff
@@ -16,65 +16,88 @@
 //  See the License for the specific language governing permissions and
 //  limitations under the License.
 
-/// The superclass for all graphs. Defined as a class instead of a protocol so that its subclasses can
-/// have some method implementations in common. You should generally use one of its two canonical subclasses,
-/// *UnweightedGraph* and *WeightedGraph*, because they offer more functionality and convenience.
-open class Graph<V: Equatable>: CustomStringConvertible, Sequence, Collection {
-    var vertices: [V] = [V]()
-    var edges: [[Edge]] = [[Edge]]() //adjacency lists
-    
+public protocol Graph: Collection, CustomStringConvertible {
+
+    associatedtype V: Equatable
+
+    var vertices: [V] { get set }
+
+    /// Adjacency lists
+    var edges: [[Edge]] { get set }
+
+    var vertexCount: Int { get }
+    var edgeCount: Int { get }
+
+    init()
+    init(vertices: [V])
+
+    func vertexAtIndex(_ index: Int) -> V
```

Comment:
If you wanted to merge this, you'd probably want to copy or move the docs from the protocol extension into here.